bento/src/main/java/com/yelp/android/bento/core/AsyncInflationBridge.kt

```diff
@@ -0,0 +1,163 @@
+        componentsAddedMap[component] = System.nanoTime()
+        coroutineScope.launch(defaultBridgeDispatcher) {
+            val inflations = (0 until component.count).map { i ->
+                async {
+                    val viewHolderType = component.getHolderType(i)
+                    val viewHolder: ComponentViewHolder<*, *> = constructViewHolder(viewHolderType)
+                    addViewHolder(viewHolder, viewHolderType)
+                    val (_, view) = BentoAsyncLayoutInflater.inflate(viewHolder, recyclerView, asyncInflaterDispatcher)
+                    viewMap[viewHolder] = view
+                }
+            }
+            lock.withLock {
+                inflations.awaitAll()
+            }
+            withContext(Dispatchers.Main) {
+                executeAddComponentCallback(component, addComponentCallback)
+            }
+        }
+    }
+
+    /**
+     * Retrieves a previously async inflated view if there is one.
+     */
+    fun getView(viewHolder: ComponentViewHolder<*, *>) = viewMap.remove(viewHolder)
+
+    /**
+     * Retrieves a previously created view holder of the passed in [viewHolderType].
+     */
+    fun getViewHolder(
+        viewHolderType: Any? // Class<out ComponentViewHolder<*, *>> doesn't work with Java.
+    ): ComponentViewHolder<*, *>? {
+        if (viewHolderMap.containsKey(viewHolderType)) {
+            viewHolderMap[viewHolderType]?.let { list ->
+                if (list.isNotEmpty()) {
+                    return list.poll()
+                }
+            }
+        }
+        return null
+    }
+
+    /**
+     * Records the time at which [component] was removed.
+     */
+    fun trackComponentRemoval(component: Component) {
+        componentsRemovedMap[component] = System.nanoTime()
+    }
+
+    private fun addViewHolder(
+        viewHolder: ComponentViewHolder<*, *>,
+        viewHolderType: Class<out ComponentViewHolder<*, *>>
+    ) {
+        val viewHolders = viewHolderMap.getOrPut(viewHolderType) {
+            ConcurrentLinkedDeque()
+        }
+        viewHolders.add(viewHolder)
+        viewHolderMap[viewHolderType] = viewHolders
+    }
+
+    /**
+     * Executes RecyclerViewComponentController's addComponent() callback on the main thread. If the
+     * component was removed during inflation, it won't add the component.
+     */
+    private fun executeAddComponentCallback(
+        component: Component,
+        addComponentCallback: () -> Unit
+    ) {
+        try {
+            val timeAddComponentRequested = componentsAddedMap[component]
+            val timeRemoveComponentRequested = componentsRemovedMap[component]
+            if (timeRemoveComponentRequested != null && timeAddComponentRequested != null) {
+                if (timeAddComponentRequested > timeRemoveComponentRequested) {
+                    addComponentCallback()
+                } // else { The component was removed while we've been inflating. }
```

Comment:
Just to make sure I understand what's happening here: This is the case where a component was added, we started inflating its views, it was removed, then added again, and then inflation finished for it so we are firing the callback?

Reply:
Almost. This only checks that we aren't adding a component that was removed while we were inflating its views.

Reply:
We could also hold a map of the inflation jobs, like

```
private val inflationJobs = ConcurrentHashMap<Component, Job>()
```

and populate it like 
```
fun asyncInflateViewsForComponent(component: Component, addComponentCallback: () -> Unit) {
    componentsAddedMap[component] = System.nanoTime()
    inflationJobs[component] = coroutineScope.launch(defaultBridgeDispatcher) {
        val inflations = (0 until component.count).map { i ->

(...)
```

And in the trackComponentRemoval, do
```
fun trackComponentRemoval(component: Component) {
    inflationJobs[component]?.cancel()
}
```

But it's definitely an improvement we can keep for later.